datafusion/expr-common/src/scalar.rs

```diff
@@ -0,0 +1,109 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::{
+    array::{Array, ArrayRef},
+    datatypes::DataType,
+};
+use datafusion_common::{exec_err, DataFusionError, Result, ScalarValue};
+
+#[derive(Clone, Debug)]
+pub struct Scalar {
+    value: ScalarValue,
+    data_type: DataType,
+}
+
+impl From<ScalarValue> for Scalar {
+    fn from(value: ScalarValue) -> Self {
+        Self {
+            data_type: value.data_type(),
+            value,
```

Comment:
This code shows that Scalar is a redundant wrapper around ScalarValue.
I understand this is transitionary. Can we capture the desired move forward in the code?
It's great if the code self-documents, but without any commentary this code self-documents itself as being redundant

Reply:
Totally understandable. As I mentioned https://github.com/apache/datafusion/pull/12536#issuecomment-2363690150 I wanted to first gauge opinions about my implementation (which proved successful as you raised a lot of interesting points to account for) before committing on some documentation prose.

Reply:
As I understood the idea, eventually the DataType of the scalar will be different than the underlying representation

I 100% agree with the sentiment that this intent should be captured in code comments as it is very much non obvious from the code

Reply:
> before committing on some documentation prose.

that's a good point
and actually, i have hard time with long documentation prose -- i am maybe a slow reader.
what i am looking for is exactly what you need -- to gauge opinions you need to understand what the code says today, but also what (today) it is supposed to say tomorrow

> As I understood the idea, eventually the DataType of the scalar will be different than the underlying representation

@alamb , on the face value it seems to be opposite to what i understood from @notfilippo 's https://github.com/apache/datafusion/pull/12536#discussion_r1769524231. I don't have strong opinion yet what scalar should or should not be, but would be great to understand the desired end state. Can you please help me with that?

Reply:
> > As I understood the idea, eventually the DataType of the scalar will be different than the underlying representation
> 
> @alamb , on the face value it seems to be opposite to what i understood from @notfilippo 's [#12536 (comment)](https://github.com/apache/datafusion/pull/12536#discussion_r1769524231). I don't have strong opinion yet what scalar should or should not be, but would be great to understand the desired end state. Can you please help me with that?

The main point of this PR is to remove variants of ScalarValue and represent them instead through the DataType contained in the Scalar type (which is what @\alamb is saying). The constraints put in place by the current constructors of Scalar allow it to only contain DataType fully compatible with the ScalarType provided (see `::from(ScalarValue)`) or cast-compatible (see `::try_from_array(...)`). Will try to make it clear via docs later today.